backend/src/utils/token.ts

```diff
@@ -0,0 +1,115 @@
+import {
+	ChatCompletionMessageParam,
+	ChatCompletionMessageToolCall,
+} from 'openai/resources/chat/completions';
+import { promptTokensEstimate, stringTokens } from 'openai-chat-tokens';
+
+import { CHAT_MODELS } from '@src/models/chat';
+import { chatGptTools } from '@src/openai';
+
+// max tokens each model can use
+const chatModelMaxTokens = {
+	[CHAT_MODELS.GPT_4_TURBO]: 128000,
+	[CHAT_MODELS.GPT_4]: 8192,
+	[CHAT_MODELS.GPT_4_0613]: 8192,
+	[CHAT_MODELS.GPT_3_5_TURBO]: 4097,
+	[CHAT_MODELS.GPT_3_5_TURBO_0613]: 4097,
+	[CHAT_MODELS.GPT_3_5_TURBO_16K]: 16385,
+	[CHAT_MODELS.GPT_3_5_TURBO_16K_0613]: 16385,
+};
+
+function countSingleToolCallTokens(toolCall: ChatCompletionMessageToolCall[]) {
+	return toolCall.reduce((acc, toolCall) => {
+		const {
+			function: { name, arguments: args },
+		} = toolCall;
+		return acc + stringTokens(name) + stringTokens(args);
+	}, 0);
+}
+
+// count total tool call in chat histroy tokens - not supported by openai-chat-tokens yet. To be removed when supported
+function countToolCallTokens(chatHistory: ChatCompletionMessageParam[]) {
+	let numToolCalls = 0;
+	let tokens = 0;
+	for (const message of chatHistory) {
+		if (message.role === 'assistant' && message.tool_calls) {
+			numToolCalls += message.tool_calls.length;
+			tokens += countSingleToolCallTokens(message.tool_calls);
+		}
+	}
+	return tokens + numToolCalls * 4; // 4 tokens per tool call
```

Comment:
particularly pernickety suggestion from me
```suggestion
        const TOKENS_PER_TOOL_CALL = 4;
	return tokens + numToolCalls * TOKENS_PER_TOOL_CALL ;
```